README.md
# jptime

jptime handle japanese time format.

## Installation

```sh
pip install jptime
```

## Usage

```py
from datetime import datetime
import jptime

# from string
jpt = jptime.from_str("平成元年三月三日")
assert jpt.to_tuple() == (4, 1, 3, 3)
assert jpt.to_datetime() == datetime(1989, 3, 3)

# from datetime
jpt = jptime.from_datetime(datetime(2019, 5, 1))
assert jpt.to_tuple() == (5, 1, 5, 1) # 令和1年5月1日
```

## Supported formats

- japanese era
  - era_symbol/yy/mm/dd (allow kanji number)
    (e.g. 昭和5年3月3日)
  - era_code + yymmdd
    (e.g. 3031123, H040323)
- christian era (delegate to dateutil.parser)
  (e.g. 19920323, 2018-12-12, 2018年8月13日)


CHANGELOG.md
# Change Log


## 1.1.0 - 2021-11-24

- Support era_code + yymmdd format (e.g. `H040323`)
- Support japanese yyyy-mm-dd format (e.g. ``2019年11月12日`)

## 1.0.0 - 2021-05-19

- Add py.typed
- Use japanese-number-converter instead of japanese-numbers-python(#6)

## 0.1.1 - 2020-03-01

Update release information in pyproject.toml.

## 0.1.0 - 2020-02-29

Initial release


test_jptime.py
from datetime import datetime

import pytest

import jptime

# Valid cases
jp_symbol_cases = [
    # normal cases
    ("明治10年3月23日", jptime.JPTime(1, 10, 3, 23)),
    ("大正10年4月2日", jptime.JPTime(2, 10, 4, 2)),
    ("昭和60年9月27日", jptime.JPTime(3, 60, 9, 27)),
    ("平成24年11月2日", jptime.JPTime(4, 24, 11, 2)),
    ("令和1年1月3日", jptime.JPTime(5, 1, 1, 3)),
    # various symbols and delimtiers
    ("昭和45年3月23日", jptime.JPTime(3, 45, 3, 23)),
    ("昭和45年03月23日", jptime.JPTime(3, 45, 3, 23)),
    ("㍼45年03月23日", jptime.JPTime(3, 45, 3, 23)),
    ("S45年03月23日", jptime.JPTime(3, 45, 3, 23)),
    ("S45.3.23", jptime.JPTime(3, 45, 3, 23)),
    # kanji cases
    ("平成三年三月二十三日", jptime.JPTime(4, 3, 3, 23)),
    ("平成元年三月二十三日", jptime.JPTime(4, 1, 3, 23)),
    ("平成元年三月二三日", jptime.JPTime(4, 1, 3, 23)),
    # corner cases
    ("明治1年1月25日", jptime.JPTime(1, 1, 1, 25)),  # Meiji begin day
    ("明治45年7月29日", jptime.JPTime(1, 45, 7, 29)),  # MEiji end day
    ("大正1年7月30日", jptime.JPTime(2, 1, 7, 30)),  # Taisho begin day
    ("大正15年12月24日", jptime.JPTime(2, 15, 12, 24)),  # Taisho end day
    ("昭和1年12月25日", jptime.JPTime(3, 1, 12, 25)),  # Showa begin day
    ("昭和64年1月7日", jptime.JPTime(3, 64, 1, 7)),  # Showa end day
    ("平成1年1月8日", jptime.JPTime(4, 1, 1, 8)),  # Heisei begin day
    ("平成31年4月30日", jptime.JPTime(4, 31, 4, 30)),  # Heisei end day
    ("令和1年5月1日", jptime.JPTime(5, 1, 5, 1)),  # Reiwa begin day
]
jp_code_cases = [
    # normal cases
    ("1100323", jptime.JPTime(1, 10, 3, 23)),
    ("2100402", jptime.JPTime(2, 10, 4, 2)),
    ("3600927", jptime.JPTime(3, 60, 9, 27)),
    ("4241102", jptime.JPTime(4, 24, 11, 2)),
    ("5010103", jptime.JPTime(5, 1, 1, 3)),
    # corner cases
    ("1010125", jptime.JPTime(1, 1, 1, 25)),  # Meiji begin day
    ("1450729", jptime.JPTime(1, 45, 7, 29)),  # MEiji end day
    ("2010730", jptime.JPTime(2, 1, 7, 30)),  # Taisho begin day
    ("2151224", jptime.JPTime(2, 15, 12, 24)),  # Taisho end day
    ("3011225", jptime.JPTime(3, 1, 12, 25)),  # Showa begin day
    ("3640107", jptime.JPTime(3, 64, 1, 7)),  # Showa end day
    ("4010108", jptime.JPTime(4, 1, 1, 8)),  # Heisei begin day
    ("4310430", jptime.JPTime(4, 31, 4, 30)),  # Heisei end day
    ("5010501", jptime.JPTime(5, 1, 5, 1)),  # Reiwa begin day
    # alphabet cases
    ("M100323", jptime.JPTime(1, 10, 3, 23)),
    ("T100402", jptime.JPTime(2, 10, 4, 2)),
    ("S600927", jptime.JPTime(3, 60, 9, 27)),
    ("H241102", jptime.JPTime(4, 24, 11, 2)),
    ("R010103", jptime.JPTime(5, 1, 1, 3)),
    ("M010125", jptime.JPTime(1, 1, 1, 25)),  # Meiji begin day
    ("M450729", jptime.JPTime(1, 45, 7, 29)),  # MEiji end day
    ("T010730", jptime.JPTime(2, 1, 7, 30)),  # Taisho begin day
    ("T151224", jptime.JPTime(2, 15, 12, 24)),  # Taisho end day
    ("S011225", jptime.JPTime(3, 1, 12, 25)),  # Showa begin day
    ("S640107", jptime.JPTime(3, 64, 1, 7)),  # Showa end day
    ("H010108", jptime.JPTime(4, 1, 1, 8)),  # Heisei begin day
    ("H310430", jptime.JPTime(4, 31, 4, 30)),  # Heisei end day
    ("R010501", jptime.JPTime(5, 1, 5, 1)),  # Reiwa begin day
]
christian_cases = [
    ("1927-9-11", jptime.JPTime(3, 2, 9, 11)),
    ("1930/03/23", jptime.JPTime(3, 5, 3, 23)),
    ("1947.09.11", jptime.JPTime(3, 22, 9, 11)),
    ("1970.3.23", jptime.JPTime(3, 45, 3, 23)),
    ("1970 3 23", jptime.JPTime(3, 45, 3, 23)),
    ("1970年3月23日", jptime.JPTime(3, 45, 3, 23)),
    ("1970 年3 月23 日", jptime.JPTime(3, 45, 3, 23)),
]
str_and_jptimes = jp_symbol_cases + jp_code_cases + christian_cases

dt_and_jptimes = [
    (datetime(1927, 9, 11), jptime.JPTime(3, 2, 9, 11)),
    (datetime(1930, 3, 23), jptime.JPTime(3, 5, 3, 23)),
    (datetime(1947, 9, 11), jptime.JPTime(3, 22, 9, 11)),
    (datetime(1970, 3, 23), jptime.JPTime(3, 45, 3, 23)),
    (datetime(1991, 3, 23), jptime.JPTime(4, 3, 3, 23)),
]


@pytest.mark.parametrize("s, jpt", jp_symbol_cases)
def test_from_jp_symbol(s, jpt):
    assert jptime._from_japanese_era_with_symbol(s) == jpt


@pytest.mark.parametrize("s, jpt", jp_code_cases)
def test_from_jp_code(s, jpt):
    assert jptime._from_japanese_era_with_code(s) == jpt


@pytest.mark.parametrize("s, jpt", str_and_jptimes)
def test_from_str(s, jpt):
    assert jptime.from_str(s) == jpt


@pytest.mark.parametrize("dt, jpt", dt_and_jptimes)
def test_from_datetime(dt, jpt):
    assert jptime.from_datetime(dt) == jpt


@pytest.mark.parametrize("dt, jpt", dt_and_jptimes)
def test_to_datetime(dt, jpt):
    assert jpt.to_datetime() == dt


# Invalid Cases
jp_symbol_invalid_cases = [
    "明治45年7月30日",  # 1day after Meiji end
    "大正15年12月25日",  # 1day after Taisho end
    "昭和64年1月8日",  # 1day after Showa end
    "平成31年5月1日",  # 1day after Heisei end
    "平成31年13月3日",  # month is out of range
    "平成30年2月29日",  # day is out of range
    "平成31年3月3日10時",  # contain time
    "[平成]1年3月3日",  # symbol is surrounded
]
jp_code_invalid_cases = [
    "0250323",  # era code is out of range
    "6250323",  # era code is out of range
    "1450730",  # 1day after Meiji end
    "2151225",  # 1day after Taisho end
    "3640108",  # 1day after Showa end
    "4310501",  # 1day after Heisei end
    "3311329",  # month is out of range
    "3310230",  # day is out of range
]
dates_cannot_parse = jp_symbol_invalid_cases + jp_code_invalid_cases


@pytest.mark.parametrize("s", jp_symbol_invalid_cases)
def test_from_jp_symbol_raises(s):
    with pytest.raises(jptime.JPTimeError):
        jptime._from_japanese_era_with_symbol(s)


@pytest.mark.parametrize("s", jp_code_invalid_cases)
def test_from_jp_code_raises(s):
    with pytest.raises(jptime.JPTimeError):
        jptime._from_japanese_era_with_code(s)


@pytest.mark.parametrize("s", dates_cannot_parse)
def test_from_str_raises(s):
    with pytest.raises(jptime.ParseError):
        jptime.from_str(s)


jptime.py
import re
import unicodedata
from datetime import datetime
from typing import Pattern, Tuple

from dateutil import parser
import japanese_numbers


__version__ = "0.1.1"


class JPTimeError(Exception):
    """Base class for all jptime exceptions."""


class ParseError(JPTimeError):
    """Cannot parse to JPTime"""


class ValidationError(JPTimeError):
    """Invalid date format."""


class JPEra:
    def __init__(self, name: str, code: int, symbol_pattern: Pattern, begin: datetime, end: datetime,) -> None:
        self.name = name
        self.code = code
        self.symbol_pattern = symbol_pattern
        self.begin = begin
        self.end = end


Meiji = JPEra("明治", 1, re.compile(r"明治|M|\u337e"), datetime(1868, 1, 25), datetime(1912, 7, 29))
Taisho = JPEra("大正", 2, re.compile(r"大正|T|\u337d"), datetime(1912, 7, 30), datetime(1926, 12, 24))
Showa = JPEra("昭和", 3, re.compile(r"昭和|S|\u337c"), datetime(1926, 12, 25), datetime(1989, 1, 7))
Heisei = JPEra("平成", 4, re.compile(r"平成|H|\u337b"), datetime(1989, 1, 8), datetime(2019, 4, 30))
Reiwa = JPEra("令和", 5, re.compile(r"令和|R|\u32ff"), datetime(2019, 5, 1), datetime.max)

ALL_ERAS = [
    Meiji,
    Taisho,
    Showa,
    Heisei,
    Reiwa,
]


class JPTime:
    def __init__(self, era_code: int, jp_year: int, month: int, day: int) -> None:
        if era_code <= 0 or era_code > len(ALL_ERAS):
            raise ValidationError(f"{era_code} is out of range.")
        self.era_code = era_code
        self.jp_year = jp_year
        self.month = month
        self.day = day
        try:
            self.jp_era = ALL_ERAS[era_code - 1]
            assert self.to_datetime() <= self.jp_era.end
        except (IndexError, AssertionError):
            raise ValidationError(f"{self} is invalid.")

    def __repr__(self) -> str:
        return "JPTime" + str(self.to_tuple())

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, JPTime):
            return NotImplemented
        return self.to_tuple() == other.to_tuple()

    def __gt__(self, other: object) -> bool:
        if not isinstance(other, JPTime):
            return NotImplemented
        return self.to_tuple() > other.to_tuple()

    def __hash__(self) -> int:
        return hash(self.to_datetime())

    def to_tuple(self) -> Tuple[int, int, int, int]:
        return self.era_code, self.jp_year, self.month, self.day

    def to_datetime(self) -> datetime:
        christian_year = self.jp_year + self.jp_era.begin.year - 1
        return datetime(christian_year, self.month, self.day)

    @classmethod
    def from_datetime(cls, dt: datetime) -> "JPTime":
        return from_datetime(dt)

    @classmethod
    def from_str(cls, s: str) -> "JPTime":
        return from_str(s)


def from_datetime(dt: datetime) -> "JPTime":
    """datetime -> JPTime"""
    for era in ALL_ERAS:
        if era.begin <= dt <= era.end:
            jp_year = dt.year - era.begin.year + 1
            return JPTime(era.code, jp_year, dt.month, dt.day)
    raise ParseError(f"Cannot convert {dt} to jp format.")


def from_str(date: str) -> "JPTime":
    """str -> JPTime

    Supported formats:
        - japanese era
            - era_symbol/yy/mm/dd (allow kanji number)
            - era_code + yymmdd
        - christian era (delegate to dateutil.parser)

    Examples:
        >>> from_str("平成３年３月２３日")
        JPTime(4, 3, 3, 23)
        >>> from_str("平成三年三月二十三日")
        JPTime(4, 3, 3, 23)
        >>> from_str("40323")
        JPTime(4, 3, 3, 23)
        >>> from_str("19910323")
        JPTime(4, 3, 3, 23)
        >>> from_str("1991-3-23")
        JPTime(4, 3, 3, 23)

    """
    normalized_date = unicodedata.normalize("NFKC", date)
    converters = (_from_japanese_era_with_symbol, _from_japanese_era_with_code, _from_christian_era)
    for converter in converters:
        try:
            return converter(normalized_date)
        except JPTimeError:
            continue
    raise ParseError(f"Cannot parse {normalized_date} to JPTime.")


def _from_japanese_era_with_symbol(s: str) -> JPTime:
    """symbol + year/month/day format -> JPTime
    e.g. 平成元年三月二十三日, 昭和5年3月23日, S22.9.11
    """
    for era in ALL_ERAS:
        if era.symbol_pattern.match(s):
            try:
                y, m, d = map(int, japanese_numbers.to_arabic_numbers(s.replace("元年", "1年")))
                return JPTime(era.code, y, m, d)
            except (ValueError, ValidationError):
                continue
    raise ParseError(f"{s} is invalid format.")


def _from_japanese_era_with_code(s: str) -> JPTime:
    """gyymmdd format -> JPTime"""
    try:
        g, yymmdd = divmod(int(s), 1000000)
        return JPTime(g, *_yymmdd2ymd(yymmdd))
    except (ValueError, IndexError):
        raise ParseError(f"{s} is invalid format.")


def _yymmdd2ymd(yymmdd: int) -> Tuple[int, int, int]:
    """yymmdd -> (year, month, day)

    Examples:
        >>> _yymmdd2ymd(321123)
        (32, 11, 23)
        >>> _yymmdd2ymd(320323)
        (32, 3, 23)
    """
    year, mmdd = divmod(yymmdd, 10000)
    month, day = divmod(mmdd, 100)
    return year, month, day


def _from_christian_era(s: str) -> JPTime:
    """christian era format -> JPTime

    using dateutil.parser
    """

    try:
        dt = parser.parse(s)
        return from_datetime(dt)
    except ValueError:
        raise ParseError(f"Cannot parse {s}.")


jptime/__init__.py
import re
import unicodedata
from datetime import datetime, date
from typing import Pattern, Tuple, Optional
from enum import Enum

from dateutil import parser
import jnc


__version__ = "1.1.0"


class JPTimeError(Exception):
    """Base class for all jptime exceptions."""


class ParseError(JPTimeError):
    """Cannot parse to JPTime"""


class ValidationError(JPTimeError):
    """Invalid date format."""


class JPEra:
    def __init__(self, name: str, code: int, symbol_pattern: Pattern, begin: datetime, end: datetime) -> None:
        self.name = name
        self.code = code
        self.symbol_pattern = symbol_pattern
        self.begin = begin
        self.end = end


class JPEraEnum(Enum):
    Meiji = JPEra("明治", 1, re.compile(r"明治|M|\u337e"), datetime(1868, 1, 25), datetime(1912, 7, 29))
    Taisho = JPEra("大正", 2, re.compile(r"大正|T|\u337d"), datetime(1912, 7, 30), datetime(1926, 12, 24))
    Showa = JPEra("昭和", 3, re.compile(r"昭和|S|\u337c"), datetime(1926, 12, 25), datetime(1989, 1, 7))
    Heisei = JPEra("平成", 4, re.compile(r"平成|H|\u337b"), datetime(1989, 1, 8), datetime(2019, 4, 30))
    Reiwa = JPEra("令和", 5, re.compile(r"令和|R|\u32ff"), datetime(2019, 5, 1), datetime.max)

    @classmethod
    def code2era(cls, code: int) -> JPEra:
        return list(JPEraEnum)[code - 1].value

    @classmethod
    def parse(cls, s: str) -> Optional[Tuple[JPEra, str]]:
        """parse string to (JPEra, rest)

        Examples:
            # >>> parsed = JPEraEnum.parse("平成３年３月２３日")
            # >>> assert parsed[0].code == 1
            # >>> assert parsed[1] "３年３月２３日")
        """
        for era_enum in cls:
            era = era_enum.value
            match = era.symbol_pattern.match(s)
            if match:
                return (era, s[match.end() :])


class JPTime:
    def __init__(self, era_code: int, jp_year: int, month: int, day: int) -> None:
        if era_code <= 0 or era_code > len(JPEraEnum):
            raise ValidationError(f"{era_code} is out of range.")
        self.era_code = era_code
        self.jp_year = jp_year
        self.month = month
        self.day = day
        try:
            self.jp_era = JPEraEnum.code2era(era_code)
            assert self.to_datetime() <= self.jp_era.end
        except (IndexError, AssertionError):
            raise ValidationError(f"{self} is invalid.")

    def __repr__(self) -> str:
        return "JPTime" + str(self.to_tuple())

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, JPTime):
            return NotImplemented
        return self.to_tuple() == other.to_tuple()

    def __gt__(self, other: object) -> bool:
        if not isinstance(other, JPTime):
            return NotImplemented
        return self.to_tuple() > other.to_tuple()

    def __hash__(self) -> int:
        return hash(self.to_datetime())

    def to_tuple(self) -> Tuple[int, int, int, int]:
        return self.era_code, self.jp_year, self.month, self.day

    def to_date(self) -> datetime:
        christian_year = self.jp_year + self.jp_era.begin.year - 1
        return date(christian_year, self.month, self.day)

    def to_datetime(self) -> datetime:
        date_ = self.to_date()
        return datetime(date_.year, date_.month, date_.day)

    @classmethod
    def from_datetime(cls, dt: datetime) -> "JPTime":
        return from_datetime(dt)

    @classmethod
    def from_str(cls, s: str) -> "JPTime":
        return from_str(s)


def from_datetime(dt: datetime) -> "JPTime":
    """datetime -> JPTime"""
    for era_enum in JPEraEnum:
        era = era_enum.value
        if era.begin <= dt <= era.end:
            jp_year = dt.year - era.begin.year + 1
            return JPTime(era.code, jp_year, dt.month, dt.day)
    raise ParseError(f"Cannot convert {dt} to jp format.")


def from_str(date_str: str) -> "JPTime":
    """str -> JPTime

    Supported formats:
        - japanese era
            - era_symbol/yy/mm/dd (allow kanji number)
            - era_code + yymmdd
        - christian era (delegate to dateutil.parser)

    Examples:
        >>> from_str("平成３年３月２３日")
        JPTime(4, 3, 3, 23)
        >>> from_str("平成三年三月二十三日")
        JPTime(4, 3, 3, 23)
        >>> from_str("4030323")
        JPTime(4, 3, 3, 23)
        >>> from_str("19910323")
        JPTime(4, 3, 3, 23)
        >>> from_str("1991-3-23")
        JPTime(4, 3, 3, 23)

    """
    normalized_date = unicodedata.normalize("NFKC", date_str)
    converters = (_from_japanese_era_with_symbol, _from_japanese_era_with_code, _from_christian_era)
    for converter in converters:
        try:
            return converter(normalized_date)
        except JPTimeError:
            continue
    raise ParseError(f"Cannot parse {normalized_date} to JPTime.")


def _parse_japanese_number(s: str) -> int:
    try:
        return int(s)
    except ValueError:
        return jnc.ja_to_arabic(s)


def _from_japanese_era_with_symbol(s: str) -> JPTime:
    """symbol + year/month/day format -> JPTime
    e.g. 平成元年三月二十三日, 昭和5年3月23日, S22.9.11
    """
    s = s.replace("元年", "一年")
    try:
        era, rest = JPEraEnum.parse(s)
        y, m, d = map(_parse_japanese_number, re.findall(r"(\d+|[〇一二三四五六七八九十]+)", rest))
        return JPTime(era.code, y, m, d)
    except (TypeError, ValueError):
        raise ParseError(f"{s} is invalid format.")


def _from_japanese_era_with_code(s: str) -> JPTime:
    """gyymmdd format -> JPTime"""
    try:
        parsed = JPEraEnum.parse(s)
        if parsed is None:
            g, yymmdd = divmod(int(s), 1000000)
        else:
            g = parsed[0].code
            yymmdd = int(parsed[1])
        return JPTime(g, *_yymmdd2ymd(yymmdd))
    except (ValueError, IndexError):
        raise ParseError(f"{s} is invalid format.")


def _yymmdd2ymd(yymmdd: int) -> Tuple[int, int, int]:
    """yymmdd -> (year, month, day)

    Examples:
        >>> _yymmdd2ymd(321123)
        (32, 11, 23)
        >>> _yymmdd2ymd(320323)
        (32, 3, 23)
    """
    year, mmdd = divmod(yymmdd, 10000)
    month, day = divmod(mmdd, 100)
    return year, month, day


def _from_christian_era(s: str) -> JPTime:
    """christian era format -> JPTime

    using dateutil.parser
    """
    s = re.sub(r"[年月日]", " ", s)
    try:
        dt = parser.parse(s)
        return from_datetime(dt)
    except ValueError:
        raise ParseError(f"Cannot parse {s}.")
